File: 202409/20.valid-parentheses.java
/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        if (s.length() % 2 != 0)
            return false;
        char[] stack = new char[s.length()];
        int top = 0;
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '[' || c == '{') {
                stack[top++] = c;
            } else {
                if (top == 0)
                    return false;
                if (c == ')' && stack[top - 1] != '(')
                    return false;
                if (c == ']' && stack[top - 1] != '[')
                    return false;
                if (c == '}' && stack[top - 1] != '{')
                    return false;
                top--;
            }
        }

        return top == 0;
    }
}
// @lc code=end

class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.isValid("()"));
        System.out.println(solution.isValid("()[]{}"));
        System.out.println(solution.isValid("(]"));
        System.out.println(solution.isValid("([)]"));
        System.out.println(solution.isValid("{[]}"));
    }
}

File: 202409/20.valid-parentheses.cpp
/*
 * @lc app=leetcode id=20 lang=cpp
 *
 * [20] Valid Parentheses
 */

#include <iostream>
#include <string>
#include <stack>

using namespace std;

// @lc code=start
class Solution
{
public:
    bool isValid(string s)
    {
        stack<char> st;
        for (auto &&c : s)
        {
            if (c == '(' || c == '[' || c == '{')
            {
                st.push(c);
            }
            else
            {
                if (st.empty())
                {
                    return false;
                }
                if (st.top() == '(' && c == ')' || st.top() == '[' && c == ']' || st.top() == '{' && c == '}')
                {
                    st.pop();
                }
                else{
                    return false;
                }
            }
        }
        return st.empty();
    }
};
// @lc code=end
